SwiftUI Settings on macOS

drucelweisse opened the pull request:
Migrated usage of a11ydescription, a11ycontainer settings view to swiftui and reuse in both VoiceOver Designer and VoiceOver Preview

Probably need to refactor some things with textRecognition offer and how insert it inside settings view and some ui polishing but it works!

## TODO
### Common
- [x] Use SegmentedStyle pickers for containers
- [x] Fix layout for traits
- [x] Redraw Canvas after any changes in Settings. Previously called to `canvas.publishArtboardChanges()` via delegate
    - [x] Delete item and press CMD-Z: should redraw Canvas to show element
- [x] Add "delete" button for an element and container's settings
    - [x] Delete an element by pressing `backspace` button on keyboard
- [x] Typing to custom actions, custom description. `@State` instead of `@Binding` helps
   - [ ] hint still slow
- [x] Add UI-tests to pipeline
- [ ] Compare performance to AppKit version
- [ ] Present payment error near button


### Preview
- [x] Add document restoration on launch for Preview
- [x] Manage keyboard focus for custom action and description
- [x] Show selected traits on settings screen. Restore FlowLayout?
- [x] Add loading state for Preview during document opening
   - [x] Add transition
- [x] Add scene restoration

### Adjustable
- [x] At least single element should be presented at adjustable options by default

<img width="1312" alt="Screenshot 2023-12-30 at 20 32 06" src="https://github.com/VODGroup/VoiceOverDesigner/assets/36012972/b987bfc9-798e-4de7-81be-bdd8e33e1c61">
<img width="1312" alt="Screenshot 2023-12-30 at 20 32 13" src="https://github.com/VODGroup/VoiceOverDesigner/assets/36012972/bb6689ad-3583-47a9-b18a-42914b10c19f">

drucelweisse committed:
feat: extract from preview to shared package and make building on mac

drucelweisse committed:
feat: wip containersettings on mac

drucelweisse committed:
Merge branch 'main' into feature/swiftui-settings-on-mac

drucelweisse committed:
feat: add combobox to macos

drucelweisse committed:
feat: insert element view

drucelweisse committed:
feat: delete unneccesary views

drucelweisse committed:
feat: add other views + basic payment view

drucelweisse committed:
feat: add textrecognitionoffer in swiftui

drucelweisse committed:
feat: delete appkit recognition offer view

drucelweisse committed:
Merge branch 'develop' into feature/swiftui-settings-on-mac

# Conflicts:
#	Shared/Package.swift
#	VoiceOver Designer/Features/Sources/Settings/SettingsStateViewController.swift
#	VoiceOver Designer/Features/Sources/Settings/States/Container/ContainerSettingsView.swift
#	VoiceOver Designer/Features/Sources/Settings/States/Container/ContainerSettingsViewController.storyboard
#	VoiceOver Designer/Features/Sources/Settings/States/Container/ContainerSettingsViewController.swift
#	VoiceOver Designer/Features/Sources/Settings/States/Element/ElementSettingsViewController.storyboard
#	VoiceOver Designer/Features/Sources/Settings/States/Element/ScrollViewController.swift

drucelweisse committed:
disable availablitiy for iOS

akaDuality committed:
Remove unused tests

akaDuality committed:
Merge branch 'feature/swiftui-settings-on-mac' of github.com:VODGroup/VoiceOverDesigner into feature/swiftui-settings-on-mac

# Conflicts:
#	VoiceOver Designer/Features/Tests/SettingsTests/ElementSettingsPresenterTests.swift

akaDuality committed:
Fix compilation for iOS

akaDuality commented:
Fix compilation, but had to fix error's alter for iOS:

```
    private func present(_ error: Error) {
#if os(macOS)
        let alert = NSAlert(error: error)
        alert.runModal()
#elseif os(iOS)
        // TODO: Show error
#endif
    }
```

Probably, we should output error's message directly in the interface, not in the alert

akaDuality commented:
How it was previously

<img width="1545" alt="Screenshot 2024-01-03 at 12 22 16" src="https://github.com/VODGroup/VoiceOverDesigner/assets/3120680/7d7b3639-81ea-4aab-88cd-b6b17c849a8b">

<img width="1545" alt="Screenshot 2024-01-03 at 12 22 10" src="https://github.com/VODGroup/VoiceOverDesigner/assets/3120680/2ba70774-8539-4438-9a74-f8aae2bd2959">
<img width="1545" alt="Screenshot 2024-01-03 at 12 21 27" src="https://github.com/VODGroup/VoiceOverDesigner/assets/3120680/9a6d2bcd-ffc9-4512-8207-8ec28f7c7061">
<img width="1545" alt="Screenshot 2024-01-03 at 12 21 31" src="https://github.com/VODGroup/VoiceOverDesigner/assets/3120680/342d713d-3827-4515-91b2-499e7988f662">

akaDuality committed:
Update visual of container

akaDuality committed:
Rework preview to macros

akaDuality committed:
Improve glouping of traits

akaDuality committed:
Improve value’s view

akaDuality committed:
Improve adjustable value

akaDuality committed:
Rework TextRecognitionComboBoxView

akaDuality committed:
Improve ComboBox

akaDuality committed:
Use recognitn result as environment variable

akaDuality committed:
Add text recognition to preview

akaDuality committed:
Return old AppKit ComboBox because menu can be open from keyboard

akaDuality committed:
Can delete an element

akaDuality committed:
Rename to launchApp()

akaDuality committed:
Show element after deletion undo

akaDuality committed:
Improve tests

akaDuality committed:
Comments

akaDuality committed:
Add default value when enable adjustable

akaDuality committed:
Update Canvas and Navigatior if an element’s settings have been updated

akaDuality committed:
Fix build for Preview

akaDuality committed:
Fix translation tests

akaDuality committed:
Union ElementSettingsEditorView

akaDuality committed:
Union ContainerSettingsEditorView

akaDuality committed:
Remove import of TextRecognition from ElementsSettings

akaDuality committed:
Merge branch 'develop' into feature/swiftui-settings-on-mac

akaDuality committed:
Fix settings for an element on iOS

akaDuality committed:
Not add empty line to adjustable options if there have been other options already

akaDuality committed:
Fix Recent windows

akaDuality committed:
Rework Traits menu for iOS on Context Menu

akaDuality committed:
Remove FlowLayout

akaDuality committed:
Improve layout on iOS

akaDuality committed:
Revert "Remove FlowLayout"

This reverts commit 85831c3f83821bb1fd38926c4265238e7ee06edb.

akaDuality committed:
Show all enabled traits in collection

akaDuality committed:
Refactor trait menu

akaDuality committed:
Add document restoration to Preview

akaDuality committed:
Close custom action’s keyboard by Done button

akaDuality committed:
Add focus managment to Custom description

akaDuality commented:
![photo_2024-01-05 20 44 02](https://github.com/VODGroup/VoiceOverDesigner/assets/3120680/21dfad05-16dd-4494-a542-51cc897ebd82)
![photo_2024-01-05 20 44 05](https://github.com/VODGroup/VoiceOverDesigner/assets/3120680/2b7d3e81-fbaa-4b79-8c54-0c5818b7e218)
![photo_2024-01-05 20 44 06](https://github.com/VODGroup/VoiceOverDesigner/assets/3120680/5ea9c950-c180-4734-a60f-3866062ec66c)

akaDuality committed:
Speed up typing
 Typing to custom actions, custom description or hint is slo
 Typing to custom actions, custom description or hint is slow
 Typing to custom actions, custom description or hint is slow

akaDuality committed:
Can delete by shortcut

akaDuality committed:
Select model after undo of deletion

akaDuality committed:
Make adjustable tests pass

akaDuality committed:
Fix typo

akaDuality committed:
Add loading state and tranisiton to Preview

akaDuality committed:
Remove custom decoding

akaDuality committed:
Hide status bar and improve transitiion animation

akaDuality committed:
Make hint transparent

akaDuality committed:
Remove translusenc from bar to fix scrolling
